sdk/java-sdk-protobuf/src/main/scala/kalix/javasdk/impl/action/ActionsImpl.scala

```diff
@@ -169,6 +178,44 @@ private[javasdk] final class ActionsImpl(_system: ActorSystem, services: Map[Str
   private def toProtocol(messageCodec: MessageCodec, sideEffects: Seq[SideEffect]): Seq[component.SideEffect] =
     sideEffects.map(asProtocol(messageCodec, _))
 
+  // re-use a single scheduled task for frequent events so that we don't fill up
+  // the timer with tasks only created as a guard to fail buggy services
+  private val reuseTimerForSeconds = 180 // a new timer every three minutes is good enough
+  private val futureDone = Future.successful(Done.done())
+  private def timeout(service: ActionService, command: ActionCommand): Future[Action.Effect[Any]] = {
+    def setUpNewTimeout(): Future[Done] = {
+      val timeout = akka.pattern.after(actionTimeout) { futureDone }
+      val now = Instant.now()
+      // Note: separate volatile read/update might mean multiple timers, but unlikely enough that it doesn't matter
+      previousTimeout = Some((now, timeout))
+      timeout
+    }
+
+    val timeout = previousTimeout match {
+      case None => setUpNewTimeout()
+      case Some((timeoutCreation, timeout))
+          if timeoutCreation.plusSeconds(reuseTimerForSeconds).isAfter(Instant.now()) =>
+        timeout
+      case Some(_) => setUpNewTimeout()
+    }
+
+    timeout.map(_ => throw timeoutErrorFor(service, command))
+  }
+
+  private def timeoutErrorFor(service: ActionService, command: ActionCommand) = {
+    val additionalDetails =
+      command.metadata match {
+        case Some(metadata) =>
+          val cloudEvent = MetadataImpl.of(metadata.entries).asCloudEvent()
+          ", " + Seq(
+            cloudEvent.subjectScala.map(s => s"subject: [$s]"),
+            cloudEvent.getScala("ce-sequence").map(s => s"sequence: [$s]")).flatten.mkString(", ")
+        case None => Map.empty
+      }
+    new TimeoutException(
+      s"Command to action [${service.actionClass.getOrElse(service.serviceName)}] method [${command.name}]${additionalDetails} did not complete within ${actionTimeout.toCoarsest}")
+  }
```

Comment:
just thinking if there is an easier way? it's pretty cheap to add and cancel short lived scheduler tasks, so wonder if we could keep the previous `Cancellable` instead and when scheduling new we always cancel previous, since we know that it was handled when processing next message.

Reply:
Rewritten like that in [009faf1](https://github.com/lightbend/kalix-jvm-sdk/pull/2383/commits/009faf1f3baf4ca57b865d24dfc920da892b214c)